### PR TITLE
[Trainer] Undo #29896

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR reverts #29896 as a better way to handle it is #30102, but that one requires a huge amount of code updates + more extensive offline discussion as it's quite a big change to something people use a lot.

Therefore, a better fix for #29790 is simply this:

```
default_collator = (
            DataCollatorWithPadding(tokenizer)
            if tokenizer is not None and isinstance(tokenizer, PreTrainedTokenizerBase)
            else default_data_collator
        )
```

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -490,12 +485,15 @@ def __init__(
         ):
             self.place_model_on_device = False
 
-        default_collator = DataCollatorWithPadding(tokenizer) if tokenizer is not None else default_data_collator
+        default_collator = (
+            DataCollatorWithPadding(tokenizer)
+            if tokenizer is not None and isinstance(tokenizer, PreTrainedTokenizerBase)
+            else default_data_collator
+        )
```
Comment (amyeroberts) [inline src/transformers/trainer.py:492]: Let's separate this out to a separate PR and keep this one as a pure revert

**Review by amyeroberts — COMMENTED** [review]
Thanks! Just a comment on separating out the revert and additional change

**Comment by NielsRogge** [comment]
Ok opened a separate PR for this

**Review by amyeroberts — APPROVED** [review]
Thanks for handling this!
